### PR TITLE
properly deprecate the current behavior of the `default_format` option

### LINKED ISSUE #1407: pint 0.18 does not allow to explicitly override "~" modifier in default_format
Hi, I see a change in the behavior from pint 0.17 to 0.18 which I think it might be a bug.

Consider the following code:

```python
from pint import UnitRegistry

ureg = UnitRegistry()
q = ureg.Quantity("1m")

print(f"implicit (before): {q}")
print(f"explicit (before): {q:d}")

ureg.default_format = "~"

print(f"implicit (after):  {q}")
print(f"explicit (after):  {q:d}")
```

With pint <=0.17 I get:
```
implicit (before): 1 meter
explicit (before): 1 meter
implicit (after):  1 m
explicit (after):  1 meter
```

With pint 0.18 I get:
```
implicit (before): 1 meter
explicit (before): 1 meter
implicit (after):  1 m
explicit (after):  1 m
```

IMHO, the behaviour of 0.17 is the expected one, since the default_format should not affect at all when a explicit format is used.

### PR DESCRIPTION
**Pull request by keewis**
As discussed in #1407, this rolls back the change in behavior of `default_format` (split the spec in `mspec` and `uspec` and decide for each part individually if the default should be used).

I also decided that this would be a good opportunity to rewrite the code such that `mspec` and `uspec` are separated as early as possible, such that fixing #1413 becomes a really small change.

- [x] Closes #1407
- [x] Executed ``pre-commit run --all-files`` with no errors
- [x] The change is fully covered by automated unit tests
- [x] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

### CODE REVIEW

**Comment by keewis** [comment]
apparently, there's a issue with installing `uncertainties==3.0.1`.

Also, I just noticed that `UnitRegistry.default_format` does not have a documentation entry, and that that's actually a property that sets class attributes on `Quantity`, `Unit`, and `Measurement`.

**Comment by jules-ch** [comment]
Yes I will make PR to resolve the error for uncertainties

**Comment by jules-ch** [comment]
Should be okay if you rebase @keewis

**Comment by keewis** [comment]
@hgrecco, this should be ready for reviewing and merging, after which we can issue the bugfix release we talked about in #1407 (maybe include #1422, too?)

**Comment by jules-ch** [comment]
Gonna look into this. but looks good so far !

**Comment by jules-ch** [comment]
@keewis  can u rebase to fix conflicts

**Comment by keewis** [comment]
@hgrecco, @jules-ch, I think this should be ready for a final review. It would probably be good to include this, as well as #1422, in the bugfix release (which we should try to release soonish).

**Comment by jules-ch** [comment]
@keewis can you just change the condition that is duplicated to use `if not mspec and default_mspec:`
Idk if ou received the notifications when I commented those lines.

**Comment by keewis** [comment]
> @keewis can you just change the condition that is duplicated to use if not mspec and default_mspec:

sure

> Idk if you received the notifications when I commented those lines.

I didn't, which is why I was confused when you posted https://github.com/hgrecco/pint/issues/1450#issuecomment-1028362822 Can you check if those comments are still pending and submit if so?

**Comment by jules-ch** [comment]
image
They are still pending, can't you see them.
I can make those changes directly if you want.

**Comment by keewis** [comment]
if it says "pending" you have to [submit the review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request#submitting-your-review)

**Comment by jules-ch** [comment]
Oh I see, I thought does kind of review comments were automatically sent. My bad then.

File: pint/formatting.py [hunk 1]
```diff
@@ -455,14 +455,20 @@ def _tothe(power):
...
+
+    flag_re = re.compile("(" + "|".join(known_flags + ["~"]) + ")")
     custom_flags = flag_re.findall(spec)
 
     return "".join(custom_flags)
 
 
 def remove_custom_flags(spec):
-    for flag in list(_FORMATTERS.keys()) + ["~"]:
+    for flag in sorted(_FORMATTERS.keys(), key=lambda k: -len(k)) + ["~"]:
```
[the remark this answers is not in the document]
Comment (keewis) [reply pint/formatting.py:471]: the sort is used to make sure we remove the longest formats first. For example, if we have `_FORMATTERS = {"cf": ..., "c": ...}` we want to make sure to remove the longer `cf` before `c`

File: pint/quantity.py [hunk 2]
```diff
@@ -360,18 +360,47 @@ def __format__(self, spec: str) -> str:
...
+                warnings.warn(
+                    (
+                        "The given format spec does not contain a unit formatter."
+                        " Falling back to the builtin defaults, but in the future"
+                        " the unit formatter specified in the `default_format`"
+                        " attribute will be used instead."
+                    ),
+                    DeprecationWarning,
+                )
+            if not uspec and default_uspec:
```
[the remark this answers is not in the document]
Comment (keewis) [reply pint/quantity.py]: indeed, I forgot to update that one. I apparently also forgot to add a test which would have caught that. This should be fixed now.

**Comment by hgrecco** [comment]
I like this PR and it is a great idea to split uspec and mspec early on

**Pull request merged**